containerd: add OCI default spec and settings

mchaker opened the pull request:
**Issue number:**

Closes #1703 

**Description of changes:**

This PR adds support for setting default OCI spec values. The default OCI spec stored at `/etc/containerd/cri-base.json` is applied to **all** containers run by containerd, unless otherwise configured (e.g. by Kubernetes).

The OCI spec can be found in detail at https://github.com/opencontainers/runtime-spec/blob/main/config.md .

This PR also adds some default settings that we want Bottlerocket to ship with: namely, a specific set of process capabilities (`sources/models/shared-defaults/oci-capabilities.toml`) and resource limits (`sources/models/shared-defaults/oci-resource-limits.toml`).

# **Testing done:**

## OCI Defaults Test Plan
​
### Manual Testing

#### Baseline
​
- [x] Baseline with Production AMI v1.11 (i.e. `ami-09d516ec16c94c0e0`)
  - [x] Get the OCI spec for a sample nginx workload container launched via kubernetes (save it for comparison)
  - [x] Run a pod that opens as many files as it can, as well as printing the container capabilities
    - [x] How many files were opened: 65533 due to [our containerd patch](https://github.com/bottlerocket-os/bottlerocket/blob/d2a03715618aa52b0b0d8c07e7f6194080abee44/packages/containerd/1001-cri-set-default-RLIMIT_NOFILE.patch).
    - [x] Bounding capabilities: {CAP_SETFCAP, CAP_NET_RAW, CAP_KILL, CAP_CHOWN, CAP_SETUID, CAP_FSETID, CAP_FOWNER, CAP_AUDIT_WRITE, CAP_MKNOD, CAP_SYS_CHROOT, CAP_DAC_OVERRIDE, CAP_SETPCAP, CAP_NET_BIND_SERVICE, CAP_SETGID} 
​
#### Test case: feature enabled and configured with default settings
​
Make sure there is no behavior change when not configuring the new settings.
​
- [x] Run modified AMI (ami-0f58fc68a762109e8, 7ad3110d) **without** setting `oci-defaults.capabilities`, `oci-defaults.resource-limits` in userdata
  - [x] Get the OCI spec for a nginx workload container: it matched v1.11
  - [x] Run a pod that opens as many files as it can, as well as printing the container capabilities
    - [x] How many files were opened: 65533 (matched v1.11)
    - [x] Bounding capabilities: {CAP_CHOWN, CAP_MKNOD, CAP_SYS_CHROOT, CAP_NET_BIND_SERVICE, CAP_SETFCAP, CAP_SETUID, CAP_FOWNER, CAP_DAC_OVERRIDE, CAP_KILL, CAP_SETGID, CAP_FSETID, CAP_SETPCAP, CAP_AUDIT_WRITE} (matched v1.11, with the exception of CAP_NET_RAW, which is [intentionally removed](https://github.com/bottlerocket-os/bottlerocket/issues/1703#issuecomment-1240834183))
- [x] Log in to the test node and verify that the cri-base.json file contains the desired settings default values
​
#### Test case: set capabilities
​
- [x] Run modified AMI (ami-01b0018dc3c2ad0de, 39407281)
    - [x] Add a capability (CAP_BPF) via apiclient
- [x] Run a pod that opens as many files as it can, as well as printing the container capabilities
    - [x] How many files were opened: 65533 (unchaged from defaults)
    - [x] Bounding capabilities: {CAP_SETFCAP, CAP_MKNOD, CAP_SETUID, **CAP_BPF**, CAP_SETGID, CAP_FSETID, CAP_AUDIT_WRITE, CAP_NET_BIND_SERVICE, CAP_DAC_OVERRIDE, CAP_CHOWN, CAP_KILL, CAP_FOWNER, CAP_SETPCAP, CAP_SYS_CHROOT} (capabilities changed accordingly: CAP_BPF was added)
- [x] Log in to the test node and verify that the cri-base.json file contains the desired settings values
​
#### Test case: set resource-limits
​
- [x] Run modified AMI (ami-01b0018dc3c2ad0de, 39407281)
    - [x] Use apiclient to change the RLIMIT_NOFILE soft value to 90000
- [x] Run a pod that opens as many files as it can, as well as printing the container capabilities
    - [x] How many files were opened: 89997 (changed according to new RLIMIT_NOFILE soft)
    - [x] Bounding capabilities: {CAP_SETPCAP, CAP_FOWNER, CAP_AUDIT_WRITE, CAP_CHOWN, CAP_SETGID, CAP_MKNOD, CAP_SYS_CHROOT, CAP_SETFCAP, CAP_KILL, CAP_NET_BIND_SERVICE, CAP_FSETID, CAP_SETUID, CAP_DAC_OVERRIDE} (unchanged from defaults)
- [x] Log in to the test node and verify that the cri-base.json file contains the desired settings values
​
#### Test case: set both capabilities and resource-limits
​
Combine both of the above: i.e. set capabilities and rlimit_nofile values
​
- [x] Run modified AMI (ami-0628b6b13d2b7c83e, 69c158f8)
    - [x] Use apiclient to change the RLIMIT_NOFILE soft value to 90000
    - [x] Use apiclient to add a capability (CAP_BPF)
- [x] Run a pod that opens as many files as it can, as well as printing the container capabilities
    - [x] How many files were opened: 89997 (changed according to new RLIMIT_NOFILE soft)
    - [x] Bounding capabilities: {CAP_KILL, CAP_SETUID, CAP_NET_BIND_SERVICE, CAP_DAC_OVERRIDE, CAP_BPF, CAP_SETPCAP, CAP_AUDIT_WRITE, CAP_MKNOD, CAP_SETFCAP, CAP_SYS_CHROOT, CAP_CHOWN, CAP_SETGID, CAP_FSETID, CAP_FOWNER} (capabilities changed accordingly: CAP_BPF was added)
- [x] Log in to the test node and verify that the cri-base.json file contains the desired settings values

#### Test all variants

Repeat above Manual Testing steps for the following variants:

- [x] aws-k8s-1.23
- [x] aws-k8s-1.23-nvidia
- [x] metal-k8s-1.24
- [x] vmware-k8s-1.23

Run abridged Manual Testing steps for the following variants:

- [ ] aws-k8s-1.21
- [ ] aws-k8s-1.21-nvidia
- [ ] aws-k8s-1.22
- [ ] aws-k8s-1.22-nvidia
- [ ] aws-k8s-1.24-nvidia
- [ ] aws-k8s-1.24
- [ ] metal-k8s-1.21
- [ ] metal-k8s-1.22
- [ ] metal-k8s-1.23
- [ ] vmware-k8s-1.21
- [ ] vmware-k8s-1.22
- [ ] vmware-k8s-1.24
​
### Run Sonobuoy Conformance
​
- [ ] Run Sonobuoy quick suite
- [ ] Run Sonobuoy full suite
​
### Test case: Upgrade Downgrade

- [x] v1.11 --> this version --> v1.11

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

mchaker commented on sources/models/shared-defaults/oci-capabilities.toml:
```diff
@@ -0,0 +1,17 @@
+[settings.oci-defaults.capabilities]
```
TODO: RENAME FILE: rename file to `oci-defaults-capabilities.toml`

mchaker replied:
change symlinks for:
- [x] aws-k8s-1.22
- [x] aws-k8s-1.22-nvidia
- [x] aws-k8s-1.24
- [x] aws-k8s-1.24-nvidia
- [x] metal-k8s-1.22
- [x] metal-k8s-1.24
- [x] vmware-k8s-1.22
- [x] vmware-k8s-1.24

mchaker commented on sources/models/shared-defaults/oci-capabilities.toml:
```diff
@@ -0,0 +1,17 @@
+[settings.oci-defaults.capabilities]
+# These values were sourced from the output of `ctr oci spec`, then
+# modified for Bottlerocket.
```
TODO: CHANGE COMMENT:
"These values represent the default capabilities in the default OCI spec for `containerd`, except for CAP_NET_RAW , which is disabled by default in Bottlerocket".

^ comment in file

comment in PR (outside this thread) explains how we got the default OCI spec values (ctr oci spec, containerd source code, etc)

mchaker commented on sources/models/shared-defaults/oci-defaults.toml:
```diff
@@ -0,0 +1,12 @@
+[settings.oci-defaults]
```
TODO: REMOVE
remove line

mchaker commented on sources/models/shared-defaults/oci-resource-limits.toml:
```diff
@@ -0,0 +1,3 @@
+[settings.oci-defaults.resource-limits.max-open-files]
```
TODO: RENAME FILE:
rename file to `oci-defaults-resource-limits.toml`

mchaker commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -0,0 +1,143 @@
+    MaxOpenFiles,
+}
+
+impl OciDefaultsResourceLimitType {
+    pub fn to_linux_string(&self) -> &'static str {
+        match self {
+            OciDefaultsResourceLimitType::MaxOpenFiles => "RLIMIT_NOFILE",
+        }
+    }
+}
```
TODO: COMMENT:
comment this function: we are leaving this implementation open for future rlimits

mchaker reviewed (commented):
remaining TODOs based on [early feedback](https://github.com/mchaker/bottlerocket/pull/83)

mchaker commented on sources/models/shared-defaults/oci-capabilities.toml:
```diff
@@ -0,0 +1,17 @@
+[settings.oci-defaults.capabilities]
+# These values were sourced from the output of `ctr oci spec`, then
+# modified for Bottlerocket.
+audit-write = true
+chown = true
+dac-override = true
+fowner = true
+fsetid = true
+kill = true
+mknod = true
+net-bind-service = true
+net-raw = false
+setgid = true
+setfcap = true
+setpcap = true
+setuid = true
+sys-chroot = true
```
How we acquired the default OCI spec values:

We ran `ctr oci spec` on a Bottlerocket node, which [printed the default](https://github.com/containerd/containerd/pull/4290) 1.0.2-dev values (latest as per https://github.com/opencontainers/runtime-spec/blob/main/specs-go/version.go )

mchaker replied:
From there, we removed CAP_NET_RAW as per the following discussion: https://github.com/bottlerocket-os/bottlerocket/issues/1703#issuecomment-1240834183

webern replied:
...but, we had to add CAP_NET_RAW back in because the AWS CNI requires it and previous versions of it do not specify the requirement in the podspec.

I am going to resolve this comment so that it's not a distraction.

mchaker commented:
Force push: resolved remaining TODOs based on early feedback.

mchaker commented:
Force push: resolved merge conflict

webern commented on sources/api/datastore/src/serialization/mod.rs:
```diff
@@ -228,6 +228,14 @@ mod test {
 
     #[test]
     fn ok_key() {
+        let serializer = MapKeySerializer::new();
+        let m = TestEnum::FooBar;
+        let res = m.serialize(&serializer).unwrap();
+        assert_eq!(res, "FooBar");
+    }
+
+    #[test]
+    fn ok_enum_key() {
```
After rebasing, you shouldn't have any changes in the datastore. Probably it is these lines that are breaking the build.

mchaker replied:
Removed these lines - thanks! :)

mchaker commented:
Force push: removed panic in helper and reset default capabilities to 1.0.2-dev defaults.

etungsten commented on sources/api/migration/migrations/v1.12.0/oci-defaults-setting/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+exclude = ["README.md"]
+
+[dependencies]
+migration-helpers = { path = "../../../migration-helpers", version = "0.1.0"}
```
Please check your IDE settings to include the newline at the end of files. This comment applies elsewhere as well.

etungsten commented on sources/api/migration/migrations/v1.12.0/oci-defaults-setting/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+[package]
```
I recommend separating out a commit for just the migration changes so it's easier maintain and to track and debug.

webern replied:
Done!

etungsten commented on README.md:
```diff
@@ -587,6 +587,94 @@ Once you opt-in to use additional OCI hooks, any new orchestrated containers wil
+
+The full list of capabilities that can be configured in Bottlerocket are:
+
+* `settings.oci-defaults.capabilities.audit-control`: process capability `CAP_AUDIT_CONTROL`
```
nit: it'd be nice to represent this as a table like the one above.

etungsten commented on README.md:
```diff
@@ -587,6 +587,94 @@ Once you opt-in to use additional OCI hooks, any new orchestrated containers wil
+
+Each of the `resource-limits` settings below contain two numeric fields: `.hard-limit` and `.soft-limit`, which are **32-bit unsigned integers**.
+
+Please see the `getrlimit` linux manpage for meanings of `.hard-limit` and `.soft-limit`.
```
```suggestion
Each of the `resource-limits` settings below contain two numeric fields: `hard-limit` and `soft-limit`, which are **32-bit unsigned integers**.

Please see the `getrlimit` linux manpage for meanings of `hard-limit` and `soft-limit`.
```

etungsten replied:
Maybe link to https://man7.org/linux/man-pages/man7/capabilities.7.html for the man page? This link is already referenced in code in the settings model crate

etungsten commented on README.md:
```diff
@@ -587,6 +587,94 @@ Once you opt-in to use additional OCI hooks, any new orchestrated containers wil
+`settings.oci-defaults.resource-limits.max-open-files.hard-limit` | 1048576
+`settings.oci-defaults.resource-limits.max-open-files.soft-limit` | 65536
+
+* `settings.oci-defaults.resource-limits.max-open-files`: resource limit `RLIMIT_NOFILE`
```
nit: seems weird for this to be here below the default values table for `max-open-files`. Since this is the description of what that is, maybe move it above the table?

mchaker replied:
I meant to mirror the capabilities section's "defaults then full list" pattern -- I forgot to add "The full list of resource limits that can be configured in Bottlerocket are"... before the listing.

Would adding that line before the (1-item) list of resource limits work? Or would it still be better to put the list of resource limits before the Bottlerocket defaults list? (My thinking that is if, in the future, ALL resource limits become configurable, but not all have default values, it would be better for both capabilities and resource-limits to have the shorter defaults section first, then the full list after that).

What do you think?

etungsten replied:
What you have now reads better. Thanks

etungsten commented on packages/containerd/containerd-cri-base-json:
```diff
@@ -0,0 +1,151 @@
+            "gid": 0
+        },
+        "cwd": "/",
+{{#if settings.oci-defaults.capabilities}}
```
A quick tip about handlebar templates, if you wrap the template with `~` like this, `{{~#if settings.oci-defaults.capabilities~}}`, you can then keep the indentation level you want. That would make reviewing these kind of changes much easier.

Example: https://github.com/bottlerocket-os/bottlerocket/blob/587f84bd30e39ceb227f0c3fd524bd66649438a5/packages/ecs-agent/ecs.config#L8-L10

mchaker replied:
So, like this?

```
        {{~#if settings.oci-defaults.capabilities~}}
        "capabilities": {
            {{~oci_defaults settings.oci-defaults.capabilities~}}
        },
        {{~/if~}}
        {{~#if settings.oci-defaults.resource-limits~}}
        "rlimits": [
            {{~oci_defaults settings.oci-defaults.resource-limits~}}
        ],
        {{~/if~}}
```

etungsten replied:
Yes. You got it

etungsten commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -0,0 +1,145 @@
+/// allowed to be modified in Bottlerocket. The full list of Linux resource limits can
+/// be found in the Linux manpages (for convenience at the man7.org link below.)
+/// Default values can be found in: sources/models/shared-defaults/oci-resource-limits.toml
+/// OCI spec: https://github.com/opencontainers/runtime-spec/blob/main/config.md#posix-process
```
Consider changing this github link to a permalink, i.e. https://github.com/opencontainers/runtime-spec/blob/c4ee7d12c742ffe806cd9350b6af3b4b19faed6f/config.md#posix-process

etungsten commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -0,0 +1,145 @@
+        assert_eq!(linux, actual_linux);
+    }
+
+    #[test]
+    fn linux_capability_strings() {
+        check_capability_strings(
+            OciDefaultsCapability::AuditControl,
+            "audit-control",
+            "CAP_AUDIT_CONTROL",
+        );
```
These small checks are fine, but ideally we can make this unit test test for all possible enum variants so as that list potentially changes, we're always covered. A quick look on how to do that led me to https://docs.rs/strum_macros/0.24.3/strum_macros/derive.EnumIter.html. The[ `strum` crate ](https://crates.io/crates/strum)seems lightweight (5.64kb) enough so I don't mind adding it as a test dependency (won't end up in the actual build if you add it as a [dev dep](https://doc.rust-lang.org/rust-by-example/testing/dev_dependencies.html)).

etungsten replied:
Ah silly me, I just realized what's being validated here. No need to do anything here.

etungsten commented on sources/models/src/modeled_types/mod.rs:
```diff
@@ -145,6 +145,18 @@ pub mod error {
         #[snafu(display("Invalid ECS duration value '{}'", input))]
         InvalidECSDurationValue { input: String },
 
+        #[snafu(display(
+            "Invalid OCI Default value '{}': {}. Validity rule: {}",
+            default_value_name,
+            provided_value,
+            validity_rule,
+        ))]
+        InvalidOciDefaults {
+            default_value_name: String,
+            provided_value: String,
+            validity_rule: String,
+        },
+
```
Maybe I'm missing something but I don't see this error ever being used.

mchaker replied:
You are correct - removed

etungsten commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -7,13 +7,18 @@ use handlebars::{
 use std::convert::TryFrom;
 use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::str::FromStr;
+use std::vec;
```
```suggestion
```

I'm pretty sure this is redundant if you're just doing this for the `Vec`

etungsten commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1346,9 +1361,188 @@ handlebars_helper!(any_enabled: |arg: Value| {
+    let oci_default_capabilities: HashMap<OciDefaultsCapability, bool> =
+        serde_json::from_value(value.clone())?;
+
+    // Only output the capabilities we support
+    let capabilities_lines: Vec<String> = oci_default_capabilities
+        .iter()
+        .filter(|(_, &capability_enabled)| capability_enabled)
+        .map(|(&capability, _)| format!("\"{}\"", capability.to_linux_string()))
+        .collect();
```
Is it OK for us to silently disregard unsupported "capabilities"? Whatever they may be?

mchaker replied:
The comment was misleading -- now fixed. What it should have been is "Output the capabilities that are enabled" (coming in the next force push)

etungsten commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1346,9 +1361,188 @@ handlebars_helper!(any_enabled: |arg: Value| {
+    let mut is_first = true;
+    for (rlimit_type, values) in oci_default_rlimits
+        .iter()
+        // Bottlerocket model values are optional, if we have no hard_limit or soft_limit, we skip
+        // the map entry.
```
We can make those fields not optional, is this something we'd like to error on during settings validation?

webern replied:
We fixed this with `#[model(add_option = false)]` which eliminates these unsightly issues.

etungsten commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1346,9 +1361,188 @@ handlebars_helper!(any_enabled: |arg: Value| {
+            .unwrap_or(values.soft_limit.unwrap_or_default());
+        let soft_limit = values
+            .soft_limit
+            .unwrap_or(values.hard_limit.unwrap_or_default());
```
I'm having a hard time wrapping my head around the reasoning behind this, can you add a comment to explain?

webern replied:
We fixed this with `#[model(add_option = false)]` which eliminates these unsightly issues.

mchaker commented:
Force push: address some comments

mchaker committed:
containerd: add OCI default spec and settings

This change adds support for setting default OCI spec values. The
default OCI spec stored at `/etc/containerd/cri-base.json` is applied
to all containers run by `containerd-cri`, unless otherwise configured
(e.g. by Kubernetes pod specs).

Co-authored-by: Mahdi Chaker <mmchaker@amazon.com>
Co-authored-by: Matthew James Briggs <brigmatt@amazon.com>

mchaker committed:
migrations: add migrations for 'settings.oci-defaults'

Co-authored-by: Mahdi Chaker <mmchaker@amazon.com>
Co-authored-by: Matthew James Briggs <brigmatt@amazon.com>

webern commented:
^ https://github.com/bottlerocket-os/bottlerocket/compare/e19d8eba27b22fedc5d253657899cd8d72d5d4ca..c633a175eadb7edd95f6fb1374de3ef16d8b62d8

break out the migration to its own commit

webern commented:
^ https://github.com/bottlerocket-os/bottlerocket/compare/c633a175eadb7edd95f6fb1374de3ef16d8b62d8..46d0035f43b35102443877f2338a763f2d8fedec

- Use `#[model(add_option = false)]` to require both `hard` and `soft` limits to be set (as is required by the spec).
- Add the OCI defaults settings to all the k8s variants.

webern commented:
^ https://github.com/bottlerocket-os/bottlerocket/compare/46d0035f43b35102443877f2338a763f2d8fedec..9d7001a575e61f0ee4cd246845f08d988a1672c4

Fix a small nit.

etungsten commented on README.md:
```diff
@@ -587,6 +587,100 @@ Once you opt-in to use additional OCI hooks, any new orchestrated containers wil
+
+Each of the `resource-limits` settings below contain two numeric fields: `hard-limit` and `soft-limit`, which are **32-bit unsigned integers**.
+
+Please see the ]`getrlimit` linux manpage](https://man7.org/linux/man-pages/man7/capabilities.7.html) for meanings of `hard-limit` and `soft-limit`.
```
```suggestion
Please see the [`getrlimit` linux manpage](https://man7.org/linux/man-pages/man7/capabilities.7.html) for meanings of `hard-limit` and `soft-limit`.
```

mchaker replied:
Thanks for catching this 😅  fixed :)

mchaker commented:
^ https://github.com/bottlerocket-os/bottlerocket/compare/9d7001a575e61f0ee4cd246845f08d988a1672c4..bf2b51bcb9b19109de89eceb060496a8b5c1f552

Force push: fix a small typo

bcressey commented on README.md:
```diff
@@ -587,6 +587,100 @@ Once you opt-in to use additional OCI hooks, any new orchestrated containers wil
+
+The **default** capabilities that are set in Bottlerocket are:
+
+setting key | setting value
```
nit: maybe
```suggestion
setting | default value
```

mchaker replied:
I agree it's clearer - accepted suggestion

bcressey commented on README.md:
```diff
@@ -587,6 +587,100 @@ Once you opt-in to use additional OCI hooks, any new orchestrated containers wil
+
+The full list of capabilities that can be configured in Bottlerocket are:
+
+setting key | corresponding linux process capability
+----- | -----
+`settings.oci-defaults.capabilities.audit-control` | `CAP_AUDIT_CONTROL`
+`settings.oci-defaults.capabilities.audit-read` | `CAP_AUDIT_READ`
+`settings.oci-defaults.capabilities.audit-write` | `CAP_AUDIT_WRITE`
+`settings.oci-defaults.capabilities.block-suspend` | `CAP_BLOCK_SUSPEND`
+`settings.oci-defaults.capabilities.bpf` | `CAP_BPF`
+`settings.oci-defaults.capabilities.checkpoint-restore` | `CAP_CHECKPOINT_RESTORE`
+`settings.oci-defaults.capabilities.chown` | `CAP_CHOWN`
```
I'd prefer to organize this as:
```
capability | setting | default value
CAP_AUDIT_CONTROL | settings.oci-defaults.capabilities.audit-control | false
```

With all the default-true ones first, followed by the others. Then we wouldn't have the duplication, and the sorting would be similar to [runc's list of capabilities](https://github.com/opencontainers/runc/blob/main/libcontainer/SPEC.md#security).

mchaker replied:
What would be preferable for `default value` in the case of capabilities we do not configure with default values?

A: The semantically correct setting of `false`
B: The syntactically correct "setting" of `(unset)`

mchaker replied:
Also, does this comment (reorganizing to `capability | setting | default value`) also apply to `rlimits` as well?
Meaning, reorganizing the `rlimits` section to be `resource limit | setting | default value`?

bcressey replied:
I would prefer something like `-` to indicate unset, so there's less visual noise.

Making the same change to the `rlimits` section would be good too.

mchaker replied:
Updated in the latest force push: https://github.com/bottlerocket-os/bottlerocket/compare/52cfccb27a6dfe636699d64696b3adf4f3346625..66c11153cc3521157c6d03a93e049fa1e88af43a

bcressey commented on README.md:
```diff
@@ -587,6 +587,100 @@ Once you opt-in to use additional OCI hooks, any new orchestrated containers wil
+`settings.oci-defaults.capabilities.syslog` | `CAP_SYSLOG`
+`settings.oci-defaults.capabilities.wake-alarm` | `CAP_WAKE_ALARM`
+
+##### OCI Defaults: Resource Limits
+
+Each of the `resource-limits` settings below contain two numeric fields: `hard-limit` and `soft-limit`, which are **32-bit unsigned integers**.
+
+Please see the [`getrlimit` linux manpage](https://man7.org/linux/man-pages/man7/capabilities.7.html) for meanings of `hard-limit` and `soft-limit`.
```
This is quite a lot of change to land in one commit. The functionality is great, no complaints there, but it makes it a bit harder for reviewers to digest.

As a suggestion for commit structure in the future, I'd recommend:
* one commit to add the scaffolding and base functionality
* one commit to add settings + template fields for rlimits (smaller)
* one commit to add settings + template fields for capabilities (larger)
* one commit for required migrations

It's fine for this PR though.

mchaker replied:
Apologies for the monolithic PR structure and thank you for explaining a better way. I'll keep this comment in my notebook as a reminder for my next PR.

bcressey commented on sources/models/src/aws-k8s-1.22-nvidia/defaults.d/95-oci-defaults.toml:
```diff
@@ -0,0 +1 @@
+../../../shared-defaults/oci-defaults.toml
```
It's a minor thing, but I'd prefer to have these in the 75-79 or 85-89 range, together with `??-oci-hooks.toml`, rather than at the end of the 90 range. That way all the OCI related defaults share the same range.

mchaker replied:
Symlinks renamed in latest force push

bcressey commented on sources/models/shared-defaults/oci-defaults-resource-limits.toml:
```diff
@@ -0,0 +1,3 @@
+[settings.oci-defaults.resource-limits.max-open-files]
+hard-limit = 1048576
+soft-limit = 65536
```
In preparation for future Docker support, we should give these files a name to keep them distinct:
* `oci-defaults-containerd-cri.toml`
* `oci-defaults-containerd-cri-resource-limits.toml`
* `oci-defaults-containerd-cri-capabilities.toml`

The capabilities will be the same but the resource limits and configuration files will be different.

mchaker replied:
Accepted suggestion, thank you

bcressey commented on sources/api/migration/migrations/v1.12.0/oci-defaults-setting/src/main.rs:
```diff
@@ -0,0 +1,25 @@
+/// `settings.oci-defaults.capabilities` and
+/// `settings.oci-defaults.resource-limits`
+fn run() -> Result<()> {
+    migrate(common_migrations::AddPrefixesMigration(vec![
+        "settings.oci-defaults",
+    ]))
```
```suggestion
    migrate(common_migrations::AddPrefixesMigration(vec![
        "settings.oci-defaults",
        "services.oci-defaults",
        "configuration-files.oci-defaults",
    ]))
```

mchaker replied:
Thank you, accepted suggestion

bcressey commented on sources/models/shared-defaults/oci-defaults.toml:
```diff
@@ -0,0 +1,10 @@
+[metadata.settings.oci-defaults]
+affected-services = ["oci-defaults", "containerd"]
```
Need a metadata migration for the affected services.

mchaker replied:
Added in the latest force push: https://github.com/bottlerocket-os/bottlerocket/compare/ec2a9f46efb263d54caee5b2df2f4d9be6c3fad2..1318d9557178011eb609a2ab2ba9165199481e83

bcressey commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -0,0 +1,143 @@
+/// Linux kernel capabilities: https://man7.org/linux/man-pages/man7/capabilities.7.html
+///
+/// Default capabilities are specified in the following file:
+/// * sources/models/shared-defaults/oci-capabilities.toml
```
I'd probably clean up all of these links and just say:
```suggestion
/// OciDefaultsCapability specifies which process capabilities are
/// allowed to be set in the default OCI spec.
```

`sources/models/shared-defaults/oci-capabilities.toml` is extremely likely to be renamed during a refactor, and this comment probably won't stay correct. The OCI spec link seems a little out-of-place here, and the man page is pretty easy to find by anyone who knows they're working with Linux capabilities.

mchaker replied:
Agreed - wouldn't want stale references in comments. Accepted suggestion

bcressey commented on sources/models/src/modeled_types/oci_defaults.rs:
```diff
@@ -0,0 +1,143 @@
+/// Linux resource limits: https://man7.org/linux/man-pages/man2/getrlimit.2.html#DESCRIPTION
+///
+/// Default resource limits are specified in the following file:
+/// * sources/models/shared-defaults/oci-resource-limits.toml
```
```suggestion
/// OciDefaultsResourceLimitType specifies which resource limits are
/// allowed to be set in the default OCI spec.
```

mchaker replied:
Accepted suggestion

mchaker commented:
^ https://github.com/bottlerocket-os/bottlerocket/compare/bf2b51bcb9b19109de89eceb060496a8b5c1f552..ec2a9f46efb263d54caee5b2df2f4d9be6c3fad2

Force push: address most PR comments

arnaldo2792 commented on sources/api/migration/migrations/v1.12.0/oci-defaults-setting/src/main.rs:
```diff
@@ -0,0 +1,27 @@
+/// `settings.oci-defaults.capabilities` and
+/// `settings.oci-defaults.resource-limits`
+fn run() -> Result<()> {
+    migrate(common_migrations::AddPrefixesMigration(vec![
+        "settings.oci-defaults",
+        "services.oci-defaults",
+        "configuration-files.oci-defaults",
+    ]))
```
Since this is a feature for k8s only, you may want to exclude the execution of the migration for all the other variants.

mchaker replied:
Done! Thanks :)

mchaker commented:
^ https://github.com/bottlerocket-os/bottlerocket/compare/ec2a9f46efb263d54caee5b2df2f4d9be6c3fad2..1318d9557178011eb609a2ab2ba9165199481e83
^ https://github.com/bottlerocket-os/bottlerocket/compare/1318d9557178011eb609a2ab2ba9165199481e83..52cfccb27a6dfe636699d64696b3adf4f3346625

Latest force pushes: fix migrations

arnaldo2792 commented on sources/Cargo.toml:
```diff
@@ -27,6 +27,8 @@ members = [
     # "api/migration/migrations/vX.Y.Z/..."
     # (all previous migrations archived; add new ones after this line)
     "api/migration/migrations/v1.12.0/k8s-private-pki-path",
+    "api/migration/migrations/v1.12.0/oci-defaults-setting",
```
Shouldn't all the migration-related changes be part of the migrations commit?

mchaker replied:
Yes. I'm not sure what happened here, I'll look into it

mchaker replied:
Yes, it is part of the migrations commit. Is it not showing that way to you?

mchaker commented:
^ https://github.com/bottlerocket-os/bottlerocket/compare/52cfccb27a6dfe636699d64696b3adf4f3346625..66c11153cc3521157c6d03a93e049fa1e88af43a

Force push: clean up README settings tables

arnaldo2792 commented on sources/api/migration/migrations/v1.12.0/oci-defaults-setting/src/main.rs:
```diff
@@ -0,0 +1,31 @@
+            "services.oci-defaults",
+            "configuration-files.oci-defaults",
+        ]))
+    } else {
+        Ok(())
+    }
```
Nit: are you forced to handle the branch? If you are not, it will be sufficient to just return `Ok(())`

mchaker replied:
The compiler complained about not including `else`

mchaker commented:
^ https://github.com/bottlerocket-os/bottlerocket/compare/66c11153cc3521157c6d03a93e049fa1e88af43a..e3d47c4f1be7227f08fa554b9c4aad6af025ae55

Force push: add metadata migration to Release.toml

etungsten reviewed (approved):
:pizza: 

LGTM assuming all the testing looks good.

webern commented:
^ rebase https://github.com/bottlerocket-os/bottlerocket/compare/e3d47c4f1be7227f08fa554b9c4aad6af025ae55..6a7d5a626706cf74155089a75f89db8170157bc7

webern commented:
^ Attempt to fix metadata migration https://github.com/bottlerocket-os/bottlerocket/compare/6a7d5a626706cf74155089a75f89db8170157bc7..8be44e04b94087399ccd2a3d6c2957558f7711e5

webern commented:
^ try even harder to fix the migrations https://github.com/bottlerocket-os/bottlerocket/compare/8be44e04b94087399ccd2a3d6c2957558f7711e5..8a04adde8601155ca1aa26389f7d522b94dde41a

stmcginnis commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1346,9 +1360,177 @@ handlebars_helper!(any_enabled: |arg: Value| {
+    let mut is_first = true;
+    for (rlimit_type, values) in &oci_default_rlimits {
+        if !is_first {
+            result_lines.push_str(",\n");
```
Just to note - I don't think it makes that much difference - but you could have `results_line` be an array, then use `result_lines.join(",\n")` at the end to get the resulting comma separated final string.

stmcginnis reviewed (approved):
Looks good to me, testing done looks good, and tests are passing. I'm not strong on the migrations, so definitely would be good to have someone that understands those better take a closer look.

bcressey commented:
Can you fix the co-authored-by tags in the commit messages?
```
Co-authored-by: name <mmchaker@amazon.com>
Co-authored-by: name <brigmatt@amazon.com>
```

Also, this description is not correct:
> The default OCI spec stored at `/etc/containerd/cri-base.json` is applied to all containers run by `containerd`, unless otherwise configured (e.g. by Kubernetes pod specs).

The changes here only affect `containerd-cri`, and not containers launched by other means (host / bootstrap containers, Docker for ECS variants).

bcressey commented on sources/api/migration/migrations/v1.12.0/oci-defaults-setting-metadata/src/main.rs:
```diff
@@ -0,0 +1,28 @@
+/// to include itself and containerd on upgrade, and to remove those values on
+/// downgrade, depending on the running variant.
+fn run() -> Result<()> {
+    if cfg!(variant_runtime = "k8s") {
+        migrate(AddMetadataMigration(&[SettingMetadata {
+            metadata: &["affected-services"],
+            setting: "settings.oci-defaults",
+        }]))?;
+    };
```
If I understand correctly - conditionalizing this on the variant is just for defensive / documentation purposes, it isn't actually necessary for `aws-ecs-1` which won't have this metadata.

bcressey replied:
nit: I don't think that last semicolon is needed:

```suggestion
    if cfg!(variant_runtime = "k8s") {
        migrate(AddMetadataMigration(&[SettingMetadata {
            metadata: &["affected-services"],
            setting: "settings.oci-defaults",
        }]))?;
    }
```

bcressey commented on sources/api/migration/migrations/v1.12.0/oci-defaults-setting/src/main.rs:
```diff
@@ -0,0 +1,31 @@
+/// `settings.oci-defaults`, which will initially contain
+/// `settings.oci-defaults.capabilities` and
+/// `settings.oci-defaults.resource-limits`
+fn run() -> Result<()> {
+    if cfg!(variant_runtime = "k8s") {
+        migrate(common_migrations::AddPrefixesMigration(vec![
+            "settings.oci-defaults",
+            "services.oci-defaults",
+            "configuration-files.oci-defaults",
+        ]))
+    } else {
+        Ok(())
+    }
+}
```
to avoid the "else" branch:
```suggestion
fn run() -> Result<()> {
    if cfg!(variant_runtime = "k8s") {
        migrate(common_migrations::AddPrefixesMigration(vec![
            "settings.oci-defaults",
            "services.oci-defaults",
            "configuration-files.oci-defaults",
        ]))?;
    }
    
    Ok(())
}
```
```

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1346,9 +1360,177 @@ handlebars_helper!(any_enabled: |arg: Value| {
+
+    let capabilities_lines_joined = capabilities_lines.join(",\n");
+
+    let capabilities_section = format!(
+        "\"bounding\": [
+{capabilities_bounding}
+],
+\"effective\": [
+{capabilities_effective}
+],
+\"permitted\": [
+{capabilities_permitted}
+]",
+        capabilities_bounding = capabilities_lines_joined,
+        capabilities_effective = capabilities_lines_joined,
+        capabilities_permitted = capabilities_lines_joined,
+    );
```
You can use a combination of raw strings and `concat!` to clean this up so the string literal isn't flush with the left margin:
```suggestion
    let capabilities_section = format!(
        concat!(
            r#""bounding": ["#,
            "{capabilities_bounding}",
            "]",
            r#""effective": ["#,
            "{capabilities_effective}",
            "]",
            r#""permitted": ["#,
            "{capabilities_permitted}",
            "]",
        ),
        capabilities_bounding = capabilities_lines_joined,
        capabilities_effective = capabilities_lines_joined,
        capabilities_permitted = capabilities_lines_joined,
    );
```

(same for the rlimit function)

mchaker replied:
This is much nicer, thank you!

bcressey reviewed (approved):
LGTM apart from a few style nits and the commit messages.

webern commented:
^ https://github.com/bottlerocket-os/bottlerocket/compare/8a04adde8601155ca1aa26389f7d522b94dde41a..684e813d7c88e27dba71a12c0c1bc7826110c141

- Avoid the else branch
- nit: I don't think that last semicolon is needed:
- You can use a combination of raw strings and concat! to clean this up so the string literal isn't flush with the left margin:
- Use Vec<String>.join
- Can you fix the co-authored-by tags in the commit messages?
- The changes here only affect containerd-cri, and not containers launched by other means (host / bootstrap containers, Docker for ECS variants).

webern commented:
^ https://github.com/bottlerocket-os/bottlerocket/compare/684e813d7c88e27dba71a12c0c1bc7826110c141..b0d09dc82fced3815b691c7ef14d5cd8f9cc3229

Fix a small oops

webern commented:
^ fix another oops https://github.com/bottlerocket-os/bottlerocket/compare/b0d09dc82fced3815b691c7ef14d5cd8f9cc3229..352a0a1477ff9084703c3be9ba8a0e1075ca1ce3

webern reviewed (approved):
I performed upgrade/downgrade testing from v1.11.0 -> v1.12.0 -> v1.11.0

On v1.11 and v1.12 I ran an ngnix pod and checked its oci spec with a command like this:

```sh
ctr -n k8s.io c info --spec 5a724a53cf
```

The specs were the same on both versions of Bottlerocket (i.e. if a customer does not use these new settings, their pods will not be affected).

:+1:
